wirego_remote(rust): add support for Wirego implemented in Rust

whiskeyo opened the pull request:
This MR is still WIP, but I am actively working to finish the implementation. Currently the `minimal` example based on the Python one, using the PCAP available in the Python example, shows the same results (at least I hope it does, the screenshot below shows it):

![image](https://github.com/user-attachments/assets/eee7f856-53db-4aae-b520-89b41c0a4f18)

TODOs:

- **DONE** add better error handling, so parsing failure does not crash the lib (if you could tell me which commands are "must haves" and should crash, and which of them can fail gracefully, it'll be great)
- **NOT VALID** add the option to disable cache (but why would we want it?)
- **CAN BE DONE LATER** add better logging with the option to enable/disable them
- **CAN BE DONE LATER / NOT VALID** add unbinding the `zeromq::RepSocket` after the main loop is finished
- **CAN BE DONE LATER / NOT VALID** add signal handling (it can probably wait, but you can tell how crucial it is)
- introduce some CI for: building the library, running tests, and releasing the package
- cleanup `Cargo.toml` from unused libraries

Limitations:

- Sadly [`zeromq`](https://crates.io/crates/zeromq) does not yet support the UDP, thus the library will not have the same functionality as you've implemented in Go or Python, but in case it is added in future - we shouldn't need lots of changes.

Once the MR is ready, I'll remove the [DRAFT] from its name :) I just need a bit of help from your side to fully understand how the Wirego Remote should work.

NothNoth commented:
Hi !

Wow, that looks really great!

To anwser your questions:

  - Error handling and crash failures: the wirego plugin (on the Wireshark side) is pretty tolerant to errors. If at some point the plugin does not respond corretly or returns an error at the end of the day Wireshark will just.. show nothing. Thus a good and simple approche is just to return an error through the ZMQ tunnel and nothing else is something goes wrong.
  - Cache enable/disable : in some use cases you may want to disable the cache. Sometimes being able to parse a whole session multiple times can be useful, this allows you to update a packet descriptor based on what you see later on the session.
  - ZMQ/UDP: I don't think that a big issue, most people will use IPC when running local and TCP when running on a remote device anyway!

NothNoth commented:
Feel free to ask any other questions and thanks again for your great work!

whiskeyo commented:
Thank you! I have some questions about your answer:

> * Error handling and crash failures: the wirego plugin (on the Wireshark side) is pretty tolerant to errors. If at some point the plugin does not respond corretly or returns an error at the end of the day Wireshark will just.. show nothing. Thus a good and simple approche is just to return an error through the ZMQ tunnel and nothing else is something goes wrong.

So sending back the message with a single frame `"\x00"` should be the right thing to do in every case, right?

> * Cache enable/disable : in some use cases you may want to disable the cache. Sometimes being able to parse a whole session multiple times can be useful, this allows you to update a packet descriptor based on what you see later on the session.

So it's allowing Wireshark to perform two pass (or multipass) run while dissecting packets?

> * ZMQ/UDP: I don't think that a big issue, most people will use IPC when running local and TCP when running on a remote device anyway!

That's great then :D

Now a few other questions from me, because I am curious about it:

1. Should there be some CI job for building and testing the wirego remotes? Examples could be also included, and since tshark is already built, we could also have some regression jobs to verify that examples are working fine.
2. Do you consider releasing the Wirego Remote in Python to PyPi and once the rust implementation is done - to cargo?
3. When creating the instance of the `WiregoListener`, can the `wirego_field_id` be anything, or it has to be incrementally raising?
4. Is there a chance of calling some internal Wireshark dissector from the plugin? I mean the embedded protocol inside another one, so I am implementing the `foo` dissector and I want to call the `bar` dissector that is already implemented in Wireshark.

NothNoth commented:
> Thank you! I have some questions about your answer:
> 
> > * Error handling and crash failures: the wirego plugin (on the Wireshark side) is pretty tolerant to errors. If at some point the plugin does not respond corretly or returns an error at the end of the day Wireshark will just.. show nothing. Thus a good and simple approche is just to return an error through the ZMQ tunnel and nothing else is something goes wrong.
> 
> So sending back the message with a single frame `"\x00"` should be the right thing to do in every case, right?


Absolutely, just reply with a 0x00 !

> > * Cache enable/disable : in some use cases you may want to disable the cache. Sometimes being able to parse a whole session multiple times can be useful, this allows you to update a packet descriptor based on what you see later on the session.
> 
> So it's allowing Wireshark to perform two pass (or multipass) run while dissecting packets?

Actually Wireshark will always perform a two pass when loading a pcap. When the pcap is loaded, all packets are processed. When you scroll on the Window, all displayed packets are re-processed so that the plugin may eventually update the results based on the complete context of the capture.
The cache mechanism is really specific to Wirego and need to be implemented on the package side.
You can take a look on the Go/Python packages to see how it's done, but basically:

  - During the process_dissect_packet, fill a map with thje packet number as key, and the dissection results as a value
  - You can use this map when you are requested for the complete dissection results (result_get_protocol, etc.)
  - If the cache is enabled, during process_dissect_packet check first if the packets has already been processed in the cache map


> > * ZMQ/UDP: I don't think that a big issue, most people will use IPC when running local and TCP when running on a remote device anyway!
> 
> That's great then :D
> 
> Now a few other questions from me, because I am curious about it:
> 
>     1. Should there be some CI job for building and testing the wirego remotes? Examples could be also included, and since tshark is already built, we could also have some regression jobs to verify that examples are working fine.

Currently there's none, but that would be cool yes!

>     2. Do you consider releasing the Wirego Remote in Python to PyPi and once the rust implementation is done - to cargo?

Absolutely, when it will be a bit more "proof-read"

>     3. When creating the instance of the `WiregoListener`, can the `wirego_field_id` be anything, or it has to be incrementally raising?

It can be anything, as long as there's no duplicate values (which would be reject by wirego)

>     4. Is there a chance of calling some internal Wireshark dissector from the plugin? I mean the embedded protocol inside another one, so I am implementing the `foo` dissector and I want to call the `bar` dissector that is already implemented in Wireshark.

Mmhh that's a good question, I'm not really sure. If I set a "protocol" value to HTTP for example, will it trigger the HTTP dissector?

NothNoth commented:
Hi @whiskeyo !

Here are a few answers to your questions/remarks:

**Part 1 :**
Sure, we can release a first version without the cache feature. It's a bonus and we can clearly do without
For the logs: can you eventually add a static flag somewhere in the code to enable/disable the logs?

**Part 2:**
That's a pretty cool feature. I do not plan to add it to Wirego for now. That's a tricky question actually: I want to keep wirego simple. There are many plugin features that are not available when using wirego. If I add them all, at the end of the day creating a wirego plugin will be quite complex and a bit intimidating.
My current position is to stick with the current set of features, unless there's a real need for something specific

**Part 3:**
That's right! Can you create some improvement issues so that we can keep track?

**Part 4:**
No problem, I'll take a look this week so that we can merge a first version of the rust package


Thanks again!

NothNoth commented:
Hi @whiskeyo !

I'm currently looking for someone at Quarkslab more fluent than me in Rust for the review :)

One additional question: I'll be talking about Wirego in a few weeks at SSTIC 25 (https://www.sstic.org/2025/news/) and Pass The Salt (https://2025.pass-the-salt.org/). Do you mind if I cite your name for the Rust contribution?

NothNoth commented on wirego_remote/rust/wirego/README.md:
```diff
@@ -0,0 +1,48 @@
+
+- `get_name` should return the protocol name, e.g. "eCPRI 2.0"
+- `get_filter` should return the protocol filter, e.g. "ecpri"
+- `get_fields` should return all fields that can be dissected by Wireshark
```
should return all fields that may be returned by your Wireshark plugin

NothNoth commented on wirego_remote/rust/wirego/examples/minimal/main.rs:
```diff
@@ -0,0 +1,151 @@
+    fn get_fields(&self) -> Vec<wirego::types::WiresharkField> {
+        vec![
+            wirego::types::WiresharkField {
+                wirego_field_id: 1,
```
Instead of using "1" here (and 2,3,4...) can you define an enum instead so that you can reuse it later during dissect?

whiskeyo commented on wirego_remote/rust/wirego/src/lib.rs:
```diff
@@ -0,0 +1,602 @@
+                println!("Response: {:?}", zmq_response);
+                send_zmq_message(&mut self.zmq_socket, zmq_response).await
+            }
+            ZmqCommandReq::SetupDetectInt(setup_detect_int) => {
```
Oh, then I had a wrong understanding, I thought that it is "take n-th filter of any type". Aligned accordingly for both `setup_detect_int` and `setup_detect_string` handling

whiskeyo commented on wirego_remote/rust/wirego/tests/integration_with_wirego_bridge.rs:
```diff
@@ -0,0 +1,655 @@
+use wirego::{Wirego, WiregoListener};
+use zeromq::{Socket, SocketRecv, SocketSend, ZmqMessage};
+
+const TEST_ENDPOINT: &str = "/tmp/test_wirego";
```
There probably is a way to do it, but this was meant to be used on UNIX only, since AFAIK - IPC is not supported on Windows. I added another integration test to verify the plugin with TCP, and the one using IPC is now running on UNIXes only.

whiskeyo commented:
Hey @NothNoth! I believe that I fixed all issues that you've pointed out. I am not sure what way of working you prefer in case of comment reviews, thus I left all of them unresolved and in some of them I added my own comments. If you agree with these changes, LMK if I should close these threads or you'll do it :) Thank you!

NothNoth commented:
That's merged!
Thanks a lot for your work, I'm super happy that we now have a Rust package for Wirego, that's a great addition!
